Support reflection-based avro serialization

bobby opened the pull request:
This pull request attempts to address #385, to enable the use of Avro without requiring IDL/JSON/etc the associated code-generation build complexity and multi-project coupling.

Serialization is pretty straightforward, so I've added a configurable boolean flag to `AbstractKafkaAvroSerDe`, and added a boolean parameter to `AvroSchemaUtils.getSchema`.

Deserialization is as bit harder, so I've done some cleanup and refactoring of the existing `AbstractKafkaAvroDeserializer` codebase in order to support the use of reflection.  Current limitations (which require some review and discussion):

* A concrete class (i.e. the class to instantiate and populate with avro record fields) must be supplied to the `ReflectionAvroDeserializer` constructor (and therefore also to the constructor of `ReflectionAvroSerde`)
* As with any `Deserializer` implementation, only a single concrete type is supported per instance.

@MichaelDrogalis @ept @benstopford @miguno: any thoughts/feedback?

bobby commented:
[clabot:check]

MichaelDrogalis commented:
Hey @bobby, this is fantastic! Thanks for contributing this!

@mageshn, this patch would greatly benefit the stream processing teams. Can someone from your team review this?

bobby commented:
@mageshn @rayokota Any feedback on this PR?  I'm happy to make any changes you think necessary.

guitcastro commented:
Please merge this ASAP, I really need it ;)

rhauch commented:
@bobby thanks for the contribution. Apologies for the delay -- we've been working hard on feature completion for the upcoming AK 2.3. One of those upstream changes affected the code that SR uses for coordination, and even though your PR doesn't affect those classes, this PR will need to be rebased to bring in some of those changes so that it compiles again.

Would you please rebase this on top of the latest `master` branch, and push to your branch to automatically update this PR? Thanks!

rhauch commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -252,4 +227,88 @@ private Schema getReaderSchema(Schema writerSchema) {
+          }
+        } catch (IOException e) {
+          throw new SerializationException("Error deserializing Avro message for id "
+              + schemaId, e);
```
The old code was catching RuntimeException in case deserialization cause a NullPointerException, etc. Do we not want to keep that behavior?

bobby commented on avro-serde/src/main/java/io/confluent/kafka/streams/serdes/avro/ReflectionAvroDeserializer.java:
```diff
@@ -0,0 +1,74 @@
+
+  @Override
+  public T deserialize(final String topic, final byte[] bytes) {
+    return inner.deserialize(topic, bytes, ReflectData.get().getSchema(type));
```
Yes, I completely agree.  I assumed that the Avro library was doing some memoization under the covers, but I didn't actually check :-).  It make sense to only do the reflection-heavy work at initialization.

bobby commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -16,29 +16,27 @@
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import kafka.utils.VerifiableProperties;
```
Yeah, my IDE got a little eager in some places.  I'll revert the non-essentials.

bobby commented:
@rhauch I've rebased and tests are passing.  Prior to rebasing, I addressed the issues you raised in your review.

Would you mind confirming that I've addressed the issues to your satisfaction?  Is there anything else that needs to be done?  Docs?

bobby commented:
@rhauch Just checking in on this.  Is there anything else that needs to be done on this PR?

bobby commented:
@rhauch Just checking in again.  Is there anything else I need to do on this PR?

rhauch commented:
I've started a rebuild; the original build had strange Jenkins failures.

mageshn commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -252,4 +231,89 @@ private Schema getReaderSchema(Schema writerSchema) {
+    private final ByteBuffer buffer;
+    private final int schemaId;
+
+    private Schema cachedSchemaFromRegistry = null;
```
I was trying to understand if there were any benefits to this cached instances. The CachedSchemaRegistryClient already caches the schema. Unless there are specific reasons, I would prefer to not cache here and simplify the code.

mageshn commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -252,4 +231,89 @@ private Schema getReaderSchema(Schema writerSchema) {
+      this.schemaId = buffer.getInt();
+    }
+
+    Schema getSchemaFromRegistry() {
```
Perhaps call it `schemaFromRegistry()`

bobby replied:
```suggestion
    Schema schemaFromRegistry() {
```

bobby replied:
@mageshn I've added a commit to address these issues.

mageshn commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -252,4 +231,89 @@ private Schema getReaderSchema(Schema writerSchema) {
+      return cachedSchemaFromRegistry;
+    }
+
+    Schema getSchemaForDeserialize() {
```
call it `schemaForDeserialize()` and we can move the actual implementation into the context class.

mageshn commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroDeserializer.java:
```diff
@@ -252,4 +231,89 @@ private Schema getReaderSchema(Schema writerSchema) {
+      return cachedSchemaForDeserialize;
+    }
+
+    String getSubject() {
```
perhaps call it `subjectName` and move the actual implementation here

mageshn commented on avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroSerDeConfig.java:
```diff
@@ -93,6 +93,11 @@
+  public static final String SCHEMA_REFLECTION_CONFIG = "schema.reflection";
+  public static final boolean SCHEMA_REFLECTION_DEFAULT = false;
+  public static final String SCHEMA_REFLECTION_DOC =
+          "If true, uses the Avro reflection API when serializing/deserializing ";
```
might be useful to specify that it can't be enabled for primitives.

mageshn commented:
Thanks a lot for incorporating the comments. More importantly, It would be great if you could help me understand the question around how a user would specify the Type for a Normal Consumer and the Serde. 
> I also had a question around how this would work when a normal KafkaConsumer needs to use the reflection for deserialize. Wouldn't the deserializer need to know the target class type? For the serde, I see that the Type is passed to the constructor. I'm not sure if Kstreams can invoke a specific constructor building the serde. May be the type should also be a configuration?

bobby commented:
@mageshn In our use of Streams and Consumers, we explicitly pass Deserializers/Serdes to the constructors.  In order to make config-driven Serdes work, we could make the Generic type a configuration parameter.  Would that satisfy both the Deserializer and Serde cases?

mageshn commented:
@bobby I think that would certainly work for config based serializers & serde.

rayokota commented:
Thanks for the PR @bobby !  I'm going to merge this for now and open a separate issue to track using this via config parameters.
